File: mini-services/excel-service/routers/daily.py
"""Ежедневный учёт API."""
from __future__ import annotations

import os
import tempfile
from typing import Optional

from fastapi import APIRouter, HTTPException, UploadFile, File, Query

from deps import run_blocking
import daily_tracking
import daily_validation
from data_paths import UPLOAD_DIR

router = APIRouter()


@router.get("/api/daily-tracking/sites")
async def daily_sites(
    active_only: bool = Query(True),
    detailed: bool = Query(False),
):
    try:
        if detailed:
            items = await run_blocking(daily_tracking.list_ploshchadki_detailed, active_only)
            return {"sites": [i["name"] for i in items], "items": items}
        return {"sites": await run_blocking(daily_tracking.list_ploshchadki, active_only)}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/api/daily-tracking")
async def daily_list(
    date: str = Query(...),
    location_id: Optional[str] = None,
    combined: bool = Query(False),
    limit: int = Query(5000),
    offset: int = Query(0),
):
    try:
        return await run_blocking(
            daily_tracking.get_rows,
            date=date,
            location_id=location_id,
            combined=combined,
            limit=limit,
            offset=offset,
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/api/daily-tracking/stats")
async def daily_stats(
    date: str = Query(...),
    location_id: Optional[str] = None,
    combined: bool = Query(False),
):
    try:
        return await run_blocking(
            daily_tracking.get_stats,
            date=date,
            location_id=location_id,
            combined=combined,
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.delete("/api/daily-tracking/data")
async def daily_clear(
    date: str = Query(...),
    location_id: Optional[str] = Query(None),
    combined: bool = Query(False),
    user_role: Optional[str] = Query(None),
    user_sites: Optional[str] = Query(None, description="Площадки пользователя через |"),
):
    sites_list = [s for s in (user_sites or "").split("|") if s.strip()]
    try:
        if combined:
            result = await run_blocking(
                daily_tracking.clear_combined_report,
                date,
                role=user_role,
                user_sites=sites_list or None,
            )
        else:
            if not location_id:
                raise HTTPException(status_code=400, detail="Укажите location_id или combined=true")
            result = await run_blocking(
                daily_tracking.clear_site_date,
                location_id,
                date,
                role=user_role,
                user_sites=sites_list or None,
            )
        if result.get("error"):
            msg = str(result["error"])
            if "Нет прав" in msg:
                raise HTTPException(status_code=403, detail=msg)
            raise HTTPException(status_code=400, detail=msg)
        return result
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/api/daily-tracking/combined/build")
async def daily_build_combined(
    date: str = Query(...),
    user_role: Optional[str] = Query(None),
):
    try:
        result = await run_blocking(
            daily_tracking.build_combined_report,
            date,
            role=user_role,
            user_sites=None,
        )
        if result.get("error"):
            raise HTTPException(status_code=400, detail=result["error"])
        return result
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/api/daily-tracking/validate")
async def daily_validate(
    date: str = Query(...),
    location_id: Optional[str] = None,
    combined: bool = Query(False),
):
    try:
        result = await run_blocking(
            daily_validation.validate_report,
            date,
            location_id=location_id,
            combined=combined,
        )
        if result.get("error"):
            raise HTTPException(status_code=400, detail=result["error"])
        return result
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/api/daily-tracking/aup")
async def daily_aup_status():
    try:
        return await run_blocking(daily_tracking.get_aup_status)
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/api/daily-tracking/aup")
async def daily_aup_upload(file: UploadFile = File(...)):
    os.makedirs(UPLOAD_DIR, exist_ok=True)
    suffix = os.path.splitext(file.filename or ".xlsx")[1] or ".xlsx"
    fd, tmp = tempfile.mkstemp(suffix=suffix, dir=UPLOAD_DIR)
    os.close(fd)
    try:
        content = await file.read()
        with open(tmp, "wb") as out:
            out.write(content)
        result = await run_blocking(
            daily_tracking.save_aup_file,
            tmp,
            file.filename or "АУП_РОП_ИТР.xlsx",
        )
        if result.get("error"):
            raise HTTPException(status_code=400, detail=result["error"])
        return result
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
    finally:
        try:
            os.remove(tmp)
        except OSError:
            pass


@router.post("/api/daily-tracking/upload")
async def daily_upload(
    file: UploadFile = File(...),
    location_id: str = Query(...),
    date: str = Query(...),
    confirm: bool = Query(False),
    user_role: Optional[str] = Query(None),
    user_sites: Optional[str] = Query(None, description="Площадки пользователя через |"),
):
    os.makedirs(UPLOAD_DIR, exist_ok=True)
    suffix = os.path.splitext(file.filename or ".xlsx")[1] or ".xlsx"
    fd, tmp = tempfile.mkstemp(suffix=suffix, dir=UPLOAD_DIR)
    os.close(fd)

    try:
        content = await file.read()
        with open(tmp, "wb") as out:
            out.write(content)

        sites_list = [s for s in (user_sites or "").split("|") if s.strip()]

        def _run():
            return daily_tracking.upload_excel(
                tmp,
                location_id=location_id,
                tracking_date=date,
                original_name=file.filename or "upload.xlsx",
                confirm=confirm,
                role=user_role,
                user_sites=sites_list or None,
            )

        result = await run_blocking(_run)
        if result.get("requiresConfirm"):
            raise HTTPException(status_code=409, detail=result)
        if result.get("error"):
            msg = str(result["error"])
            if "Нет прав" in msg:
                raise HTTPException(status_code=403, detail=msg)
            raise HTTPException(status_code=400, detail=msg)
        return result
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
    finally:
        try:
            os.remove(tmp)
        except OSError:
            pass

File: src/hooks/excel-api/daily.ts
import type { createHealthApi } from '@/hooks/excel-api/health'
import type { AppUser } from '@/lib/app-auth'

type HealthApi = ReturnType<typeof createHealthApi>

export type DailySiteItem = {
  name: string
  opStatus: 'active' | 'paused' | 'finished'
  statusLabel: string
  custom?: boolean
}

function authFields(user: AppUser | null) {
  if (!user) return {}
  return {
    user_role: user.role,
    user_sites: user.sites.join('|'),
  }
}

export function createDailyApi({ ensureExcelBackend }: HealthApi) {
  const dailySites = async (activeOnly = true, detailed = false) => {
    await ensureExcelBackend({ lenient: true })
    const sp = new URLSearchParams({
      action: 'sites',
      active_only: String(activeOnly),
    })
    if (detailed) sp.set('detailed', 'true')
    const res = await fetch(`/api/excel/daily-tracking?${sp}`)
    const data = await res.json().catch(() => ({}))
    if (!res.ok) throw new Error(String((data as { detail?: string }).detail || 'Ошибка списка площадок'))
    return data as { sites: string[]; items?: DailySiteItem[] }
  }

  const dailyList = async (params: {
    date: string
    locationId?: string
    combined?: boolean
    limit?: number
    offset?: number
  }) => {
    await ensureExcelBackend({ lenient: true })
    const sp = new URLSearchParams({ action: 'list', date: params.date })
    if (params.locationId) sp.set('location_id', params.locationId)
    if (params.combined) sp.set('combined', 'true')
    sp.set('limit', String(params.limit ?? 5000))
    sp.set('offset', String(params.offset ?? 0))
    const res = await fetch(`/api/excel/daily-tracking?${sp}`)
    const data = await res.json().catch(() => ({}))
    if (!res.ok) throw new Error(String((data as { detail?: string }).detail || 'Ошибка загрузки данных'))
    return data as {
      data: Record<string, unknown>[]
      total: number
      hasCombined?: boolean
    }
  }

  const dailyStats = async (params: { date: string; locationId?: string; combined?: boolean }) => {
    await ensureExcelBackend({ lenient: true })
    const sp = new URLSearchParams({ action: 'stats', date: params.date })
    if (params.locationId) sp.set('location_id', params.locationId)
    if (params.combined) sp.set('combined', 'true')
    const res = await fetch(`/api/excel/daily-tracking?${sp}`)
    const data = await res.json().catch(() => ({}))
    if (!res.ok) throw new Error(String((data as { detail?: string }).detail || 'Ошибка статистики'))
    return data
  }

  const dailyUpload = async (params: {
    file: File
    locationId: string
    date: string
    confirm?: boolean
    replaceSiteDate?: boolean
    user?: AppUser | null
  }) => {
    await ensureExcelBackend()
    const form = new FormData()
    form.append('file', params.file)
    form.append('location_id', params.locationId)
    form.append('date', params.date)
    form.append('confirm', String(Boolean(params.confirm)))
    form.append('replace_site_date', String(Boolean(params.replaceSiteDate)))
    const auth = authFields(params.user ?? null)
    if (auth.user_role) form.append('user_role', auth.user_role)
    if (auth.user_sites) form.append('user_sites', auth.user_sites)
    const res = await fetch('/api/excel/daily-tracking', { method: 'POST', body: form })
    const data = await res.json().catch(() => ({}))
    if (res.status === 409) {
      const err = new Error('Требуется подтверждение замены') as Error & {
        requiresConfirm?: boolean
        existingCount?: number
      }
      err.requiresConfirm = true
      const detail = (data as { detail?: { existingCount?: number } }).detail
      err.existingCount = detail?.existingCount
      throw err
    }
    if (!res.ok) {
      throw new Error(
        typeof (data as { detail?: unknown }).detail === 'string'
          ? (data as { detail: string }).detail
          : (data as { error?: string }).error || 'Ошибка загрузки Excel',
      )
    }
    return data as { rowCount?: number; locationId?: string }
  }

  const dailyAddSite = async (name: string, user: AppUser) => {
    await ensureExcelBackend()
    const sp = new URLSearchParams({ user_role: user.role })
    const res = await fetch(`/api/excel/daily-tracking/sites?${sp}`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ name }),
    })
    const data = await res.json().catch(() => ({}))
    if (!res.ok) throw new Error(String((data as { detail?: string }).detail || 'Не удалось добавить площадку'))
    return data as { sites?: DailySiteItem[] }
  }

  const dailyTemplateStatus = async () => {
    await ensureExcelBackend({ lenient: true })
    const res = await fetch('/api/excel/daily-tracking/template')
    const data = await res.json().catch(() => ({}))
    if (!res.ok) throw new Error(String((data as { detail?: string }).detail || 'Ошибка шаблона'))
    return data as {
      hasTemplate?: boolean
      originalName?: string
      uploadedAt?: string
      size?: number
    }
  }

  const dailyUploadTemplate = async (file: File) => {
    await ensureExcelBackend()
    const form = new FormData()
    form.append('file', file)
    const res = await fetch('/api/excel/daily-tracking/template', { method: 'POST', body: form })
    const data = await res.json().catch(() => ({}))
    if (!res.ok) throw new Error(String((data as { detail?: string }).detail || 'Ошибка загрузки шаблона'))
    return data
  }

  const dailyClear = async (params: {
    date: string
    locationId?: string
    combined?: boolean
    user?: AppUser | null
  }) => {
    await ensureExcelBackend()
    const sp = new URLSearchParams({ date: params.date })
    if (params.combined) sp.set('combined', 'true')
    else if (params.locationId) sp.set('location_id', params.locationId)
    const auth = authFields(params.user ?? null)
    if (auth.user_role) sp.set('user_role', auth.user_role)
    if (auth.user_sites) sp.set('user_sites', auth.user_sites)
    const res = await fetch(`/api/excel/daily-tracking?${sp}`, { method: 'DELETE' })
    const data = await res.json().catch(() => ({}))
    if (!res.ok) {
      throw new Error(
        typeof (data as { detail?: unknown }).detail === 'string'
          ? (data as { detail: string }).detail
          : (data as { error?: string }).error || 'Ошибка очистки данных',
      )
    }
    return data as {
      deletedRows?: number
      deletedUploads?: number
      deletedSites?: number
      combined?: boolean
    }
  }

  const dailyBuildCombined = async (date: string, user: AppUser) => {
    await ensureExcelBackend()
    const sp = new URLSearchParams({ date, user_role: user.role })
    const res = await fetch(`/api/excel/daily-tracking/combined?${sp}`, { method: 'POST' })
    const data = await res.json().catch(() => ({}))
    if (!res.ok) throw new Error(String((data as { detail?: string }).detail || 'Ошибка формирования «Общий»'))
    return data as { rowCount?: number; hasCombined?: boolean }
  }

  const dailyValidate = async (params: {
    date: string
    locationId?: string
    combined?: boolean
  }) => {
    await ensureExcelBackend({ lenient: true })
    const sp = new URLSearchParams({ date: params.date })
    if (params.locationId) sp.set('location_id', params.locationId)
    if (params.combined) sp.set('combined', 'true')
    const res = await fetch(`/api/excel/daily-tracking/validate?${sp}`)
    const data = await res.json().catch(() => ({}))
    if (!res.ok) throw new Error(String((data as { detail?: string }).detail || 'Ошибка проверки'))
    return data as {
      errors: Array<{
        check: string
        row: number
        locationId?: string
        tabNumber?: string
        fio?: string
        field?: string
        message: string
        mainDb?: Record<string, unknown>
      }>
      errorCount: number
      hasErrors: boolean
      rowCount: number
    }
  }

  const dailyAupStatus = async () => {
    await ensureExcelBackend({ lenient: true })
    const res = await fetch('/api/excel/daily-tracking/aup')
    const data = await res.json().catch(() => ({}))
    if (!res.ok) throw new Error(String((data as { detail?: string }).detail || 'Ошибка АУП'))
    return data as {
      hasAup?: boolean
      originalName?: string
      uploadedAt?: string
      size?: number
      mappingCount?: number
    }
  }

  const dailyUploadAup = async (file: File) => {
    await ensureExcelBackend()
    const form = new FormData()
    form.append('file', file)
    const res = await fetch('/api/excel/daily-tracking/aup', { method: 'POST', body: form })
    const data = await res.json().catch(() => ({}))
    if (!res.ok) throw new Error(String((data as { detail?: string }).detail || 'Ошибка загрузки АУП'))
    return data
  }

  const dailyExport = async (params: {
    date: string
    locationId?: string
    combined?: boolean
    fileName?: string
  }) => {
    await ensureExcelBackend()
    const sp = new URLSearchParams({ date: params.date })
    if (params.locationId) sp.set('location_id', params.locationId)
    if (params.combined) sp.set('combined', 'true')
    const res = await fetch(`/api/excel/daily-tracking/export?${sp}`)
    if (!res.ok) {
      const data = await res.json().catch(() => ({}))
      throw new Error(String((data as { detail?: string }).detail || 'Ошибка выгрузки'))
    }
    const blob = await res.blob()
    const url = URL.createObjectURL(blob)
    const a = document.createElement('a')
    a.href = url
    a.download = params.fileName || `Ежедневный_учет_${params.date}.xlsx`
    a.click()
    URL.revokeObjectURL(url)
  }

  return {
    dailySites,
    dailyList,
    dailyStats,
    dailyUpload,
    dailyAddSite,
    dailyTemplateStatus,
    dailyUploadTemplate,
    dailyExport,
    dailyClear,
    dailyBuildCombined,
    dailyValidate,
    dailyAupStatus,
    dailyUploadAup,
  }
}
